File: atcoder/abc159/e.py
import itertools

H, W, K = map(int, input().split())
A = [[int(x) for x in input()] for _ in range(H)]


def solve(blocks):
    n_vert_cut = 0
    n_block = len(blocks)
    vals = [0 for _ in range(n_block)]

    for c in range(W):
        adds = [block[c] for block in blocks]
        if any(add > K for add in adds):
            return H * W

        sums = [val + add for val, add in zip(vals, adds)]
        if any(x > K for x in sums):
            n_vert_cut += 1
            vals = adds
        else:
            vals = sums

    return n_vert_cut + n_block - 1


ans = H * W
for mask in itertools.product([0, 1], repeat=H - 1):
    mask = [1] + list(mask) + [1]
    pivots = [r for r in range(H + 1) if mask[r]]
    blocks = [A[p1:p2] for p1, p2 in zip(pivots[:-1], pivots[1:])]
    blocks = [[sum(row[c] for row in block) for c in range(W)] for block in blocks]
    # print(blocks, '->', solve(blocks))
    ans = min(ans, solve(blocks))
print(ans)

File: atcoder/abc159/e.cpp
#include <bits/stdc++.h>
using namespace std;

void print_vector(const vector<int>& v) {
    for (int x : v) {
        cout << x << " ";
    }
    cout << endl;
}

int H, W, K;
string A[10];

int check(const vector<int>& pivots) {
    // cout << "pivots: ";
    // print_vector(pivots);

    int cut_cnt = 0;
    int n_block = pivots.size() - 1;
    auto val = vector<int>(n_block, 0);

    for (int c = 0; c < W; c++) {
        // cout << "col " << c << ":" << endl;
        vector<int> add(n_block, 0);
        for (int block_id = 0; block_id < n_block; block_id++) {
            int block_start = pivots[block_id];
            int block_end = pivots[block_id + 1];
            // cout << "block " << block_start << "," << block_end << endl;
            for (int r = block_start; r < block_end; r++) {
                add[block_id] += A[r][c] - '0';
            }
        }

        // cout << "val:";
        // print_vector(val);
        // cout << "add:";
        // print_vector(add);

        for (int block_id = 0; block_id < n_block; block_id++) {
            if (add[block_id] > K) {
                // cout << "impossible" << endl;
                return H * W + 100;
            }
        }

        bool cut = false;
        for (int block_id = 0; block_id < n_block; block_id++) {
            if (val[block_id] + add[block_id] > K) {
                cut = true;
                cut_cnt++;
                break;
            }
        }

        for (int block_id = 0; block_id < n_block; block_id++) {
            if (cut) {
                val[block_id] = add[block_id];
            }
            else {
                val[block_id] = val[block_id] + add[block_id];
            }
        }
    }
    // cout << "cut_cnt:" << cut_cnt << endl;

    return cut_cnt + n_block - 1;
}

int solve() {
    int ans = H * W + 100;
    for (int mask = 0; mask < (1 << (H + 1)); mask++) {
        if ((mask & (1 << 0)) == 0) continue;
        if ((mask & (1 << H)) == 0) continue;
        auto pivots = vector<int>();
        for (int r = 0; r <= H; r++) {
            if (mask & (1 << r)) {
                pivots.push_back(r);
            }
        }

        ans = min(ans, check(pivots));
    }

    return ans;
}

int main() {
    ios::sync_with_stdio(false);

    cin >> H >> W >> K;
    for (int i = 0; i < H; i++) {
        cin >> A[i];
    }

    cout << solve() << endl;

    return 0;
}